### DISCUSSION TITLE
Is there a way to specify a zAttr cdftype in pycdf?

### DISCUSSION THREAD

**Discussion by Alrobbertz** [spacepy/spacepy#715] [Support and Q&A] [answered]
I'd like to specify the cdftype used when setting a zAttr when creating a CDF file. It looks it is 'guessed' based on the type of the `Var` looking at the docs here:
https://spacepy.github.io/autosummary/spacepy.pycdf.zAttrList.html#spacepy.pycdf.zAttrList 

Is there a better way to do this? If not are you open to contributions to allow this? 

### Minimal example to reproduce issue:

```
from spacepy.pycdf import CDF

c = CDF("test.cdf", create=True)

c["my_var"] = [1.0, 2.0, 3.0]
print(f"Variable Type: {type(c['my_var'][...][0])}")

test_int_attr_value = 4
print(f"Attr Initial (Python) Type: {type(test_int_attr_value)}")
c["my_var"].meta["test_int_attr"] = test_int_attr_value

print(f"Attr Type in CDF: {type(c['my_var'].meta['test_int_attr'])}")```

```
Outputs:
```
Variable Type: <class 'numpy.float32'>
Attr Initial (Python) Type: <class 'int'>
Attr Type in CDF: <class 'numpy.float32'>
```
I'd like the "Attr Type in CDF" to be something like `const.CDF_INT8`. 

### OS, Python version, and dependency version information:
```
macOS-13.5-arm64-arm-64bit
sys.version_info(major=3, minor=8, micro=16, releaselevel='final', serial=0)
numpy=1.24.2
scipy=1.10.1
matplotlib=3.7.1
spacepy=0.4.1 # Downloaded from pip
```

**Comment by jtniehof** [comment, marked as answer]
(converted to a discussion)

You can specify the attribute type with the [new method](https://spacepy.github.io/autosummary/spacepy.pycdf.AttrList.html#spacepy.pycdf.AttrList.new) of AttrList (common to zAttrList and gAttrList):

```python
import spacepy.pycdf

c = spacepy.pycdf.CDF('test.cdf', create=True)
c['my_var'] = [1., 2, 3]
c['my_var'].attrs.new('test_int_attr', 4, spacepy.pycdf.const.CDF_INT8)
print(c['my_var'].attrs)
# test_int_attr: 4 [CDF_INT8]
```

There's a lot hiding in "the type is chosen to match the data". In particular, if `data` is of a numpy type, it is matched exactly:
```python
import numpy
import spacepy.pycdf

c = spacepy.pycdf.CDF('test.cdf', create=True)
c['my_var'] = [1., 2, 3]
c['my_var'].attrs['test_int_attr'] = numpy.int64(4)
print(c['my_var'].attrs)
# test_int_attr: 4 [CDF_INT8]
```

Or:
```python
import numpy
import spacepy.pycdf

c = spacepy.pycdf.CDF('test.cdf', create=True)
c['my_var'] = spacepy.dmarray(
    [1., 2, 3],
    attrs={'test_int_attr': numpy.int64(4)})
print(c['my_var'].attrs)
# test_int_attr: 4 [CDF_INT8]
```

As an aside, it seems unlikely that INT8 is the best option; CDF is Fortran-like and specifies length in bytes, whereas numpy specifies in bits. So you might mean INT1?

It's fairly rare for [ISTP-compliant metadata](https://spdf.gsfc.nasa.gov/istp_guide/vattributes.html) to have a numerical zEntry that doesn't match the CDF type of the corresponding zVariable, which is why this is the default.
